### PR TITLE
TYP: Signature of "rename" incompatible with supertype "NDFrame"

### LINKED ISSUE #40977: TYP: Signature of "rename" incompatible with supertype "NDFrame"
This isn't currently detected as `Series.rename` is untyped and so is unchecked

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [x] closes #40977
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

---

@WillAyd this should still be backwards compatible, it just makes the signature compatible with its parent's

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -1698,7 +1699,7 @@ def size(self) -> FrameOrSeriesUnion:
             result = self._obj_1d_constructor(result)
 
         if not self.as_index:
-            result = result.rename("size").reset_index()
+            result = cast(Series, result.rename("size")).reset_index()
```
Comment (MarcoGorelli) [inline pandas/core/groupby/groupby.py:1702]: previously, the revealed type of `result.rename('size')` was `Any`, while now it's `Series | None`. the cast could be avoided with overloads, but that's a separate issue

Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py:1702]: if a cast is needed, i think better to use a type: ignore to indicate that we can try to annotate _obj_1d_constructor (or self.as_index?) more effectively

File: pandas/tests/series/methods/test_rename.py [hunk 2]
```diff
@@ -101,3 +102,16 @@ def test_rename_callable(self):
...
+        s = Series([1, 2])
+        with pytest.raises(TypeError, match="Cannot specify both 'mapper' and 'index'"):
+            s.rename(str, index=str)
+
+    def test_rename_none(self):
+        # GH 40977
+        s = Series([1, 2], name="foo")
+        result = s.rename(None)
+        expected = Series([1, 2])
+        tm.assert_series_equal(result, expected)
```
Comment (jbrockmendel) [inline pandas/tests/series/methods/test_rename.py:117]: this works on master doesnt it?

**Comment by jbrockmendel** [comment]
can we remove rewrite_axis_style_signature from DataFrame.rename?

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/groupby.py]: good point, thanks!

File: pandas/tests/series/methods/test_rename.py (hunk 2 above)
Comment (MarcoGorelli) [reply pandas/tests/series/methods/test_rename.py:117]: yup, was just adding an extra check to make sure this doesn't break anything

**Comment by MarcoGorelli** [comment]
> can we remove rewrite_axis_style_signature from DataFrame.rename?

I haven't yet made sense fo what `rewrite_axis_style_signature` does, but I"ll look into it

File: pandas/core/series.py [hunk 3]
```diff
@@ -4355,14 +4355,16 @@ def align(
 
     def rename(
         self,
-        index=None,
+        mapper=None,
         *,
+        index=None,
```
Comment (jreback) [inline pandas/core/series.py:4360]: umm, you are adding args?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/methods/test_rename.py [hunk 4]
```diff
@@ -101,3 +102,16 @@ def test_rename_callable(self):
         tm.assert_series_equal(result, expected)
 
         assert result.name == expected.name
+
+    def test_rename_method_and_index(self):
+        # GH 40977
+        s = Series([1, 2])
```
Comment (jbrockmendel) [inline pandas/tests/series/methods/test_rename.py:108]: nitpick: can we avoid 1-letter variable names?

File: pandas/core/series.py (hunk 3 above)
Comment (MarcoGorelli) [reply pandas/core/series.py:4360]: Yes, but it's just for compatibility with the signature in `generic.py`, from a user perspective everything should be the same

File: pandas/tests/series/methods/test_rename.py (hunk 4 above)
Comment (MarcoGorelli) [reply pandas/tests/series/methods/test_rename.py]: sure, thanks (might be worth adding such a rule to pandas' flake8 plugin)

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Comment by jreback** [comment]
still working?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
needs rebase

File: pandas/core/series.py (hunk 3 above)
Comment (MarcoGorelli) [reply pandas/core/series.py:4475]: Yes - the idea is that now it's called `mapper`, but that if a user passes `index`, then it'll still work as before (for backwards compatibility).
And if someone passes both `mapper` and `index`, it'll raise an error

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @MarcoGorelli i realize now why this is convoluted, will be glad to remove this in 2.0

**Pull request merged**